test/icinga-checkresult.cpp

```diff
@@ -809,4 +814,183 @@ BOOST_AUTO_TEST_CASE(service_flapping_ok_over_bad_into_ok)
 
 #endif /* I2_DEBUG */
 }
+
+BOOST_AUTO_TEST_CASE(suppressed_notification)
+{
+	/* Tests that suppressed notifications on a Checkable are sent after the suppression ends if and only if the first
+	 * hard state after the suppression is different from the last hard state before the suppression. The test works
+	 * by bringing a service in a defined hard state, creating a downtime, performing some state changes, removing the
+	 * downtime, bringing the service into another defined hard state (if not already) and checking the requested
+	 * notifications.
+	 */
+
+	const std::vector<ServiceState> states {ServiceOK, ServiceWarning, ServiceCritical, ServiceUnknown};
+
+	for (bool isVolatile : {false, true}) {
+		for (int checkAttempts : {1, 2}) {
+			for (ServiceState initialState: states) {
+				for (auto s1 : states) for (auto s2 : states) for (auto s3 : states) for (auto s4 : states) {
+					const std::vector<ServiceState> sequence {s1, s2, s3, s4};
+
+					std::string testcase;
+					{
+						std::ostringstream buf;
+						buf << "volatile=" << isVolatile
+							<< " checkAttempts=" << checkAttempts
+							<< " sequence={" << Service::StateToString(initialState);
+						for (ServiceState s: sequence) {
+							buf << " " << Service::StateToString(s);
+						}
+						buf << "}";
+						testcase = buf.str();
+					}
+					std::cout << "Test case: " << testcase << std::endl;
+
+					// Create host and service for the test.
+					Host::Ptr host = new Host();
+					host->SetName("suppressed_notifications");
+					host->Register();
+
+					Service::Ptr service = new Service();
+					service->SetHostName(host->GetName());
+					service->SetName("service");
+					service->SetActive(true);
+					service->SetVolatile(isVolatile);
+					service->SetMaxCheckAttempts(checkAttempts);
+					service->Activate();
+					service->SetAuthority(true);
+					service->Register();
+
+					dynamic_pointer_cast<ConfigObject>(host)->OnAllConfigLoaded();
+					dynamic_pointer_cast<ConfigObject>(service)->OnAllConfigLoaded();
```

Comment:
Don't say the dynamic lookup is actually needed?

Reply:
`OnAllConfigLoaded()` is `protected` in `Host` and `Service`, so I can't use it without changing the visibility. But maybe that's the better idea here, there's really no point in having them protected if they can be called like that.

Reply:
I mean: couldn't you cast e.g. (C style)?

Reply:
Other casts will probably work as well. Do you see an issue with just making the functions public in the children? (But if it's fine for you, please don't merge yet, I've just noticed that there's another instance of this for the downtime.)

Reply:
Yes, it is.